Draw wxTextCtrl focus ring natively on Mac

vslavik opened the pull request:
`NSTextView` doesn't display focus ring by default, which is why wxOSX did draw it manually, but this behavior can be overriden since OS X 10.3 with `NSView.focusRingType` property.

The HITheme-based rendering suffered from a number of non-nativeness issues:
1. didn't respect macOS 10.14+ accent colors
2. not animated as the native focus ring
3. subtly different shape of the outline
4. _noticeably_ different outline shape on macOS 11

The PR fixes this by removing `NeedsFocusRect()` and associated workaround for manually drawing focus ring inside `NSTextView` (i.e. multiline text controls) and replacing it with a way to override default focus ring type. This private interface was only used for wxTextCtrl and nothing else, so this shouldn't have any impact elsewhere.

vslavik committed:
Draw wxTextCtrl focus ring natively on Mac

NSTextView doesn't display focus ring by default, which is why wxOSX
did draw it manually, but this behavior can be overriden since OS X
10.3 with NSView.focusRingType property.

The HITheme-based rendering suffered from a number of non-nativeness
issues:
- didn't respect macOS 10.14+ accent colors
- not animated as the native focus ring
- subtly different shape of the outline
- noticeably different outline shape on macOS 11

Remove NeedsFocusRect() and associated workaround for manually drawing
focus ring inside NSTextView (i.e. multiline text controls). This
private interface was only used for wxTextCtrl and nothing else, so
this shouldn't have any impact elsewhere.

csomor commented:
Thanks a lot, I'm fine with these changes. 

The only thing which I wonder is whether we should have the triplet of non-default-exterior exposed as well, I can't test right now, whether there's a difference between default and exterior, I thought maybe the former would be suitable for grids etc...

vadz commented:
Great, thanks for fixing this!

Not directly related to this PR, but I wonder if we should disable focus ring automatically for the borderless (`wxNO_BORDER`) text controls? Having Mac-specific code in `wxGridCellTextEditor::DoCreate()` is really not great...

vslavik commented:
> The only thing which I wonder is whether we should have the triplet of non-default-exterior exposed as well, I can't test right now, whether there's a difference between default and exterior,

As far as I can determine `NSFocusRingTypeDefault` ("The default focus ring type for NSView or NSCell." in docs) simply means "use whatever is appropriate for this NSView", i.e. maps to one of the other two visually; there isn't any visually different kind, just `NSFocusRingTypeExterior` ("The standard Aqua focus ring.") or nothing.

IOW, the mapping in wx API is:

`NSFocusRingTypeDefault` — do not call `SetFocusRectEnabled()` at all
`NSFocusRingTypeNone` — call `SetFocusRectEnabled(false)`
`NSFocusRingTypeExterior` — call `SetFocusRectEnabled(true)`

I think that is sufficient and doesn't warrant introducing another private enum type.

---

Also, I think this by @vadz is relevant to this PR if I change its title to "Improve focus ring handling on Mac" ;), so let me address it in depth:

> Not directly related to this PR, but I wonder if we should disable focus ring automatically for the borderless (wxNO_BORDER) text controls?

I don't think so. I had a vaguely uneasy feeling about that at first (not understanding all the implications), but I have a concrete reason now: consider borderless image buttons. They don't draw a border, but they *do* draw focus ring in macOS when "Full Keyboard Access" is enabled in *System Preferences → Keyboard → Shortcuts*.

> Having Mac-specific code in wxGridCellTextEditor::DoCreate() is really not great...

Note that this is an issue only because wxOSX *changes system default* unconditionally when creating `wxTextCtrl`, and `wxGridCellTextEditor::DoCreate()` just undoes it. If wx sticked to the platform default, which is that multiline text views don't have a focus ring (they indicate focus by showing a cursor), this platform-specific code wouldn't need to be there.

I did wonder — and Vadim's comment convinces me it is more important then I realized at first — if it would make sense to:

1. have public `wxWindow::OSXSetFocusRectEnabled(bool)` as a more accessible interface
2. alternatively, have a style flag for this
3. change `wxTextCtrl` default to *not* unconditionally enable focus ring on multi-line controls

In the end, whether a text control should have focus ring is application-specific judgement call. Some should (e.g. a commit message field), some shouldn't (the only textview in a small modal window, a large editing area, a read-only text control). Apple clearly felt that *usually* you don't want it, hence AppKit's default being off. 

Any of the three above would improve `wxGridCellTextEditor` situation: 3. would remove the need entirely (but I fear would be viewed as compatibility-breaking), 1. would change it into cleaner Mac-only code, 2. would remove #ifdefs (but I think is overkill for something as obscure as this -- if you're worrying about focus ring, you're making Mac-specific tweaking and can as well use an #ifdef).

vadz commented:
I didn't even realize focus rings were not used by default for multiline text controls, but it seems perfectly clear to me that we shouldn't override the system default behaviour, and especially not unconditionally, in this case, i.e. we should do (3) from above. I don't think we need to worry about compatibility that much here, the program UI still remains perfectly usable and not showing the focus ring should be an improvement in more cases than not.

However we should also do (1), i.e. provide a way to turn the focus ring on if needed. Ideally would be to have some more "semantic" function for this, e.g. `ShowAsImportant()` or something like this, which could potentially be used in the other ports in the future (but for now would just enable focus ring under Mac and do nothing elsewhere).

csomor commented:
In early versions of MacOS text fields didn't but lists did have a focus ring, later all controls receiving keyboard input got one, then there was a time where parts of the system were using it and others not, and then it disappeared mostly, sometimes also for single line entries. So I'm for (1) and (3), (2) would be nice for declaring interfaces
I'm also wondering whether (3) could depend on a app-wide system setting  

It's not so much a function of whether the field is very important, but how the user interacts with the app. So I'm for keeping the naming close to (keyboard) 'focus' be it focus-rect or focus-ring ..,

csomor commented:
> The only thing which I wonder is whether we should have the triplet of non-default-exterior exposed as well, I can't test right now, whether there's a difference between default and exterior,
> 
> As far as I can determine NSFocusRingTypeDefault ("The default focus ring type for NSView or NSCell." in docs) simply means "use whatever is appropriate for this NSView", i.e. maps to one of the other two visually; there isn't any visually different kind, just NSFocusRingTypeExterior ("The standard Aqua focus ring.") or nothing.
> IOW, the mapping in wx API is:
> NSFocusRingTypeDefault — do not call SetFocusRectEnabled() at all
> NSFocusRingTypeNone — call SetFocusRectEnabled(false)
> NSFocusRingTypeExterior — call SetFocusRectEnabled(true)
> I think that is sufficient and doesn't warrant introducing another private enum type.

if that's the meaning of the Default Type, then I completely agree.

vadz commented:
I hope you can come up with some good name for the new function, I really don't have enough experience with macOS UI to be able to formulate what does the presence of the focus ring mean. If there is no semantic meaning I guess we could use something along the lines of `SetVisibleFocus()` or `SetOstentatiousFocus()` or `EmphasizeFocus()`...

vslavik commented:
> I really don't have enough experience with macOS UI to be able to formulate what does the presence of the focus ring mean.

I think I somewhat do, as a user, and I struggle to formulate any rule. Honestly, I think it is futile.

In fact, I struggle to find *any* native use of multiline textview with a focus ring, except in one 3rd party application. I thought Finder had it in its Info window, but that's a line-wrapped single-line control. And I'm so used to seeing it in wx that I didn't realize until now how very rare it is. 

> If there is no semantic meaning I guess we could use something along the lines of SetVisibleFocus() or SetOstentatiousFocus() or EmphasizeFocus()…

I also don't think it makes sense to try to generalize this when no other platform behaves like this — and even macOS almost doesn't. If we invent generalized terminology for something that's not ever going to be used elsewhere, realistically speaking, we'd just be obfuscating it. At least "focus rect" is something a person that knows they want to do this is able to search for.

So I really think that `wxWindow::OSXSetFocusRingEnabled(bool)` (probably better than "rect", to stick to AppKit terminology) is the best: it's the least work, least invasive to the API, most clear in its (very limited) purpose. And the main reason to even *have* it probably is in case somebody complains about the change, because the more I think about it, the more sure I'm that not even *I* will need/want to use it…

vadz commented:
I think we're still probably going to need to _disable_ focus ring for some single-line controls (unless this is not done at all under macOS?), so I think it would make sense to have a portable (and not `OSX`-prefixed) function to do it, even if it's called `EnableFocusRing()` (I'd rather have this than `Set...Enabled()`).

We also will almost certainly want to have XRC support for it.

vslavik committed:
Don’t add non-default focus ring to NSTextView

NSTextView does not natively show a focus ring by default, and it is
extremely rare thing to do in native applications.

Don't do it in wxTextCtrl either.

vslavik committed:
Allow configuring visible focus on Mac

Add wxWindow::EnableVisibleFocus() for changing focus ring behavior on
macOS (currently not implemented elsewhere, although GTK+ has a
discouraged option to do it).

vslavik commented:
> I think we're still probably going to need to disable focus ring for some single-line controls (unless this is not done at all under macOS?),

I've seen it done *once*, but I doubt it was an intentional thought-out decision, and not an artifact of using some custom controls: in some search fields in Xcode (anywhere else in the same context focus ring is shown). 

> so I think it would make sense to have a portable (and not OSX-prefixed) function to do it, even if it's called EnableFocusRing() (I'd rather have this than Set...Enabled()).

I added it per your wishes. I used `EnableVisibleFocus()` name, which is by far the best of the ones suggested here — clear, explanatory, easy to search for (more so than "focus ring" unless you're AppKit expert).

I also force-pushed a modification of the first commit to use "FocusRing" instead of "FocusRect" in macOS-native code for consistency with the native API it calls, and removed the non-native default in 9c5abea as discussed above.

> We also will almost certainly want to have XRC support for it.

I'm not going to do that myself, sorry, this PR escalated enough as it is. See the docs comment in 9f66b03 explains why — focus ring tweaking is an obscure, rarely needed thing. Compared to it  e.g. `OSXEnableAutomaticQuoteSubstitution()` is much more useful and doesn't have XRC support either.

IOW, I think this is in good shape now?

vadz commented:
Thanks, I like `EnableVisibleFocus()` name!

FWIW I think automatic quote substitution etc should have XRC support too, but if you think this one is used so rarely, we may wait until someone asks for it before adding, so I'm perfectly fine with merging this right now.

@csomor do you have any objections? If not, I'll merge it soon.

csomor commented:
@vslavik thank's a lot
